java-checks/src/main/java/org/sonar/java/checks/CollectionMethodsWithLinearComplexityCheck.java

```diff
@@ -0,0 +1,142 @@
+ * SonarQube Java
+ * Copyright (C) 2012-2017 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.checks;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+
+import org.sonar.check.Rule;
+import org.sonar.java.checks.helpers.MethodsHelper;
+import org.sonar.java.matcher.MethodMatcher;
+import org.sonar.java.matcher.TypeCriteria;
+import org.sonar.plugins.java.api.IssuableSubscriptionVisitor;
+import org.sonar.plugins.java.api.semantic.Symbol;
+import org.sonar.plugins.java.api.semantic.Type;
+import org.sonar.plugins.java.api.tree.ExpressionTree;
+import org.sonar.plugins.java.api.tree.IdentifierTree;
+import org.sonar.plugins.java.api.tree.MemberSelectExpressionTree;
+import org.sonar.plugins.java.api.tree.MethodInvocationTree;
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.plugins.java.api.tree.VariableTree;
+
+import javax.annotation.CheckForNull;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+@Rule(key = "S2250")
+public class CollectionMethodsWithLinearComplexityCheck extends IssuableSubscriptionVisitor {
+
+  private static final String JAVA_LANG_OBJECT = "java.lang.Object";
+  private static final String ARRAY_LIST = "java.util.ArrayList";
+  private static final String LINKED_LIST = "java.util.LinkedList";
+  private static final String COW_ARRAY_LIST = "java.util.concurrent.CopyOnWriteArrayList";
+  private static final String COW_ARRAY_SET = "java.util.concurrent.CopyOnWriteArraySet";
+  private static final String CONCURRENT_LINKED_QUEUE = "java.util.concurrent.ConcurrentLinkedQueue";
+  private static final String CONCURRENT_LINKED_DEQUE = "java.util.concurrent.ConcurrentLinkedDeque";
+
+  private static MethodMatcher collectionMethodMatcher() {
+    return MethodMatcher.create().typeDefinition(TypeCriteria.subtypeOf("java.util.Collection"));
+  }
+
+  private static final Map<MethodMatcher, Set<String>> matcherActualTypeMap;
+  static {
+    ImmutableMap.Builder<MethodMatcher, Set<String>> builder = ImmutableMap.builder();
+
+    MethodMatcher collectionContains = collectionMethodMatcher().name("contains").addParameter(JAVA_LANG_OBJECT);
```

Comment:
I want to target only that specific signature, why we should be less precise, when we can be exact?